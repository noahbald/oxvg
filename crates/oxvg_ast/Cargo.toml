[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
