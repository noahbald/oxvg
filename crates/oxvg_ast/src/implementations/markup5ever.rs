//! Parsing methods using xml5ever
use std::{
    cell::{Cell, RefCell},
    collections::HashSet,
};

use xml5ever::{
    driver::{parse_document, XmlParseOpts},
    interface::{NodeOrText, QuirksMode, TreeSink},
    tendril::TendrilSink,
    tree_builder::ElemName,
};

use crate::node::Node as _;

use super::shared::{Arena, Attribute, Node, NodeData, QualName, Ref};

/// parse an xml file using xml5ever as the parser.
///
/// # Errors
///
/// If the file cannot be read or parsed
pub fn parse_file<'arena>(
    source: &std::path::Path,
    arena: Arena<'arena>,
) -> Result<Ref<'arena>, std::io::Error> {
    parse_document(Sink::new(arena), XmlParseOpts::default())
        .from_utf8()
        .from_file(source)
}

/// parse an xml document using xml5ever as the parser.
pub fn parse<'arena>(source: &str, arena: Arena<'arena>) -> Ref<'arena> {
    parse_document(Sink::new(arena), XmlParseOpts::default()).one(source)
}

struct Sink<'arena> {
    arena: Arena<'arena>,
    document: Ref<'arena>,
    mode: Cell<QuirksMode>,
    line: Cell<u64>,
}

impl<'arena> Sink<'arena> {
    fn new(arena: Arena<'arena>) -> Self {
        Self {
            arena,
            document: arena.alloc(Node::new(NodeData::Document)),
            mode: Cell::new(QuirksMode::NoQuirks),
            line: Cell::new(1),
        }
    }
}

impl ElemName for &QualName {
    fn ns(&self) -> &xml5ever::Namespace {
        &self.ns
    }

    fn local_name(&self) -> &xml5ever::LocalName {
        &self.local
    }
}

impl<'arena> Sink<'arena> {
    fn new_node(&self, data: NodeData) -> &'arena mut Node<'arena> {
        self.arena.alloc(Node::new(data))
    }
}

impl<'arena> TreeSink for Sink<'arena> {
    type Handle = Ref<'arena>;
    type Output = Ref<'arena>;
    type ElemName<'a>
        = &'a QualName
    where
        Self: 'a;

    fn finish(self) -> Self::Output {
        self.document
    }

    fn parse_error(&self, _msg: std::borrow::Cow<'static, str>) {}

    fn get_document(&self) -> Self::Handle {
        self.document
    }

    fn set_quirks_mode(&self, mode: xml5ever::interface::QuirksMode) {
        self.mode.set(mode);
    }

    fn set_current_line(&self, line: u64) {
        self.line.set(line);
    }

    fn same_node(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
        x.ptr_eq(y)
    }

    fn elem_name<'a>(&'a self, target: &'a Self::Handle) -> Self::ElemName<'a> {
        match target.node_data {
            NodeData::Element { ref name, .. } => name,
            _ => panic!("not an element!"),
        }
    }

    fn get_template_contents(&self, target: &Self::Handle) -> Self::Handle {
        target
    }

    fn is_mathml_annotation_xml_integration_point(&self, handle: &Self::Handle) -> bool {
        let NodeData::Element { ref name, .. } = handle.node_data else {
            panic!("not an element!");
        };
        name.prefix.is_none() && matches!(name.local.as_ref(), "mi" | "mo" | "mn" | "ms" | "mtext")
    }

    fn create_element(
        &self,
        name: xml5ever::QualName,
        attrs: Vec<xml5ever::Attribute>,
        _flags: xml5ever::interface::ElementFlags,
    ) -> Self::Handle {
        self.new_node(NodeData::Element {
            name: QualName {
                prefix: name.prefix,
                ns: name.ns,
                local: name.local,
            },
            attrs: RefCell::new(
                attrs
                    .into_iter()
                    .map(|attr| Attribute {
                        name: QualName {
                            prefix: attr.name.prefix,
                            ns: attr.name.ns,
                            local: attr.name.local,
                        },
                        value: attr.value,
                    })
                    .collect(),
            ),
            #[cfg(feature = "selectors")]
            selector_flags: Cell::new(None),
        })
    }

    fn create_comment(&self, text: tendril::StrTendril) -> Self::Handle {
        self.new_node(NodeData::Comment(RefCell::new(Some(text))))
    }

    fn create_pi(&self, target: tendril::StrTendril, data: tendril::StrTendril) -> Self::Handle {
        self.new_node(NodeData::PI {
            target,
            value: RefCell::new(Some(data)),
        })
    }

    fn append(&self, parent: &Self::Handle, child: xml5ever::interface::NodeOrText<Self::Handle>) {
        match child {
            NodeOrText::AppendNode(node) => parent.append_child(node),
            NodeOrText::AppendText(text) => {
                parent.append_child(self.new_node(NodeData::Text(RefCell::new(Some(text)))));
            }
        }
    }

    fn append_before_sibling(&self, sibling: &Self::Handle, new_node: NodeOrText<Self::Handle>) {
        let new_node = match new_node {
            NodeOrText::AppendNode(node) => node,
            NodeOrText::AppendText(text) => self.new_node(NodeData::Text(RefCell::new(Some(text)))),
        };
        let before_sibling = sibling.previous_sibling.replace(Some(new_node));
        let parent = sibling.parent.get();

        new_node.next_sibling.set(Some(sibling));
        new_node.parent.set(parent);
        if let Some(before_sibling) = before_sibling {
            before_sibling.next_sibling.set(Some(new_node));
        } else if let Some(parent) = parent {
            parent.first_child.set(Some(new_node));
        }
    }

    fn append_based_on_parent_node(
        &self,
        element: &Self::Handle,
        prev_element: &Self::Handle,
        child: NodeOrText<Self::Handle>,
    ) {
        if element.parent.get().is_some() {
            self.append_before_sibling(element, child);
        } else {
            self.append(prev_element, child);
        }
    }

    fn append_doctype_to_document(
        &self,
        _name: tendril::StrTendril,
        _public_id: tendril::StrTendril,
        _system_id: tendril::StrTendril,
    ) {
        // doctype not needed in svg documents
    }

    fn add_attrs_if_missing(&self, target: &Self::Handle, new_attrs: Vec<xml5ever::Attribute>) {
        let NodeData::Element { ref attrs, .. } = target.node_data else {
            panic!("not an element!");
        };
        let mut attrs = attrs.borrow_mut();

        let existing_names = attrs
            .iter()
            .map(|attr| attr.name.clone())
            .collect::<HashSet<_>>();
        attrs.extend(
            new_attrs
                .into_iter()
                .map(|attr| Attribute {
                    name: QualName {
                        prefix: attr.name.prefix,
                        ns: attr.name.ns,
                        local: attr.name.local,
                    },
                    value: attr.value,
                })
                .filter(|attr| existing_names.contains(&attr.name)),
        );
    }

    fn remove_from_parent(&self, target: &Self::Handle) {
        target.remove();
    }

    fn reparent_children(&self, node: &Self::Handle, new_parent: &Self::Handle) {
        let mut current = node.first_child.take();
        let old_last_child = new_parent.last_child.take();
        if let Some(current) = current {
            if let Some(old_last_child) = old_last_child {
                old_last_child.next_sibling.set(Some(current));
                current.previous_sibling.set(Some(old_last_child));
            } else {
                debug_assert!(new_parent.first_child.get().is_none());
                new_parent.first_child.set(Some(current));
            }
        } else {
            return;
        }

        while let Some(child) = current {
            child.parent.set(Some(new_parent));
            current = child.next_sibling.get();
        }
        new_parent.last_child.set(current);
    }
}

#[test]
fn parse_markup5ever() {}
