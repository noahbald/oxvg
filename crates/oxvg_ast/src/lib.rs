#[macro_use]
extern crate bitflags;

#[cfg(feature = "style")]
#[macro_use]
extern crate smallvec;

pub mod atom;
pub mod attribute;
pub mod class_list;
pub mod document;
pub mod element;
pub mod implementations;
pub mod name;
pub mod node;

#[cfg(feature = "visitor")]
pub mod visitor;

#[cfg(feature = "parse")]
pub mod parse;

#[cfg(feature = "serialize")]
pub mod serialize;

#[cfg(feature = "selectors")]
pub mod selectors;

#[cfg(feature = "style")]
pub mod style;
