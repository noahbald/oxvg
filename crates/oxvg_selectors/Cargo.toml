[package]
name = "oxvg_selectors"
version = "0.1.0"
edition = "2021"
authors.workspace = true

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
cssparser = { workspace = true }
derivative = { workspace = true }
lazy_static = { workspace = true }
markup5ever = { workspace = true }
rcdom = { workspace = true }
selectors = "0.25.0"
traversal = "0.1.2"
xml5ever = { workspace = true }

[dev-dependencies]
oxvg_ast = { workspace = true }
