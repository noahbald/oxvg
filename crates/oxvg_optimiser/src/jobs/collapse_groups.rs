use std::cell::RefCell;

use lightningcss::{
    properties::PropertyId,
    stylesheet::{ParserOptions, StyleAttribute},
    vendor_prefix::VendorPrefix,
};
use markup5ever::{local_name, Attribute};
use oxvg_ast::{node, Attributes};
use oxvg_selectors::{
    collections::{ElementGroup, Group, INHERITABLE_ATTRS},
    Element, ElementData,
};
use serde::Deserialize;

use crate::{Job, PrepareOutcome};

#[derive(Deserialize, Default, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CollapseGroups(bool);

impl Job for CollapseGroups {
    fn prepare(&mut self, _document: &rcdom::RcDom) -> crate::PrepareOutcome {
        if self.0 {
            PrepareOutcome::None
        } else {
            PrepareOutcome::Skip
        }
    }

    fn run(&self, node: &std::rc::Rc<rcdom::Node>) {
        use rcdom::NodeData::Element as ElementData;

        let ElementData { attrs, .. } = &node.data else {
            return;
        };
        let element = Element::new(node.clone());
        let Some(parent) = element.get_parent() else {
            return;
        };

        if element.is_root() || parent.get_name() == Some(local_name!("switch")) {
            return;
        }
        if element.get_name() != Some(local_name!("g")) || node.children.borrow().len() == 0 {
            return;
        }

        move_attributes_to_child(&element, attrs);
        flatten_when_all_attributes_moved(&element, attrs);
    }
}

fn move_attributes_to_child(element: &Element, attrs: &RefCell<Vec<Attribute>>) {
    dbg!("collapse_groups: move_attributes_to_child");
    if attrs.borrow().len() == 0 {
        dbg!("collapse_groups: not moving attrs: no attrs to move");
        return;
    }

    let children: Vec<_> = element.children().collect();
    if children.len() != 1 {
        dbg!("collapse_groups: not moving attrs: many children");
        return;
    }

    let child = children.first().expect("No child after checking length");
    let ElementData {
        attrs: child_attrs, ..
    } = &child.data();

    if dbg!(is_group_identifiable(element, child))
        || dbg!(is_position_visually_unstable(element, child))
        || dbg!(is_node_with_filter(element))
    {
        dbg!("collapse_groups: not moving attrs: see true condition");
        return;
    };

    let mut new_node_attrs = Attributes::default();
    let mut new_child_attrs = Attributes::from(&*child_attrs.borrow());
    for attr in attrs.borrow().iter() {
        let Attribute { name, value } = attr.clone();
        let name = node::QualName(name.clone());

        if has_animated_attr(child, &name) {
            dbg!("collapse_groups: canelled moves: has animated_attr");
            return;
        }

        let Some(old_value) = new_child_attrs.get(&name) else {
            dbg!("collapse_groups: moved {}: same as parent", &name.0.local);
            new_child_attrs.insert(name, value);
            continue;
        };

        if name.0.local == local_name!("transform") {
            dbg!("collapse_groups: moved transform: is transform");
            new_child_attrs.insert(name, format!("{value} {old_value}").into());
        } else if old_value.to_string() == *"inherit" {
            dbg!(
                "collapse_groups: moved {}: is explicit inherit",
                &name.0.local
            );
            new_child_attrs.insert(name, value);
        } else if INHERITABLE_ATTRS.contains(name.0.local.as_ref())
            && new_child_attrs.get(&name).is_none()
        {
            dbg!("collapse_groups: canelled moves: inheritable attr is applicable");
            return;
        } else if INHERITABLE_ATTRS.contains(name.0.local.as_ref())
            || new_child_attrs.get(&name) == Some(&attr.value)
        {
            dbg!("collapse_groups: removing {}: inheritable attr is not inherited");
        } else {
            dbg!(
                "collapse_groups: ignoring {}: fails to meet movement criteria",
                &name.0.local,
            );
            new_node_attrs.insert(name, value);
        }
    }

    attrs.replace(new_node_attrs.into());
    let attrs = child.attrs();
    attrs.replace(
        new_child_attrs
            .iter()
            .map(|(name, value)| Attribute {
                name: name.0.clone(),
                value: value.clone(),
            })
            .collect(),
    );
}

fn flatten_when_all_attributes_moved(element: &Element, attrs: &RefCell<Vec<Attribute>>) {
    if attrs.borrow().len() > 0 {
        return;
    }

    let animation_group = oxvg_selectors::collections::Group::set(&ElementGroup::Animation);
    {
        let children: Vec<_> = element.children().collect();
        if children.iter().any(|child| {
            child
                .get_name()
                .is_some_and(|name| animation_group.contains(name.to_string().as_str()))
        }) {
            return;
        }
    }

    element.flatten();
}

fn has_animated_attr(node: &Element, name: &node::QualName) -> bool {
    let Some(node_name) = node.get_name() else {
        return false;
    };
    if Group::set(&ElementGroup::Animation).contains(node_name.as_ref())
        && node
            .get_attr(&local_name!("attributeName"))
            .is_some_and(|attr| attr.value.to_string() == name.to_string())
    {
        return true;
    }
    node.node
        .children
        .borrow()
        .iter()
        .any(|child| has_animated_attr(&Element::new(child.clone()), name))
}

fn is_group_identifiable(node: &Element, child: &Element) -> bool {
    child.get_attr(&local_name!("id")).is_some()
        && (node.get_attr(&local_name!("class")).is_none()
            || child.get_attr(&local_name!("class")).is_none())
}

fn is_position_visually_unstable(node: &Element, child: &Element) -> bool {
    let is_node_clipping = node.get_attr(&local_name!("clip-path")).is_some()
        || node.get_attr(&local_name!("mask")).is_some();
    let is_child_transformed_group = child.get_name() == Some(local_name!("g"))
        && child.get_attr(&local_name!("transform")).is_some();
    dbg!(is_node_clipping) || dbg!(is_child_transformed_group)
}

fn is_node_with_filter(node: &Element) -> bool {
    node.get_attr(&local_name!("filter")).is_some()
        || node.get_attr(&local_name!("style")).is_some_and(|code| {
            StyleAttribute::parse(code.value.as_ref(), ParserOptions::default()).is_ok_and(
                |style| {
                    dbg!(&style.declarations);
                    style
                        .declarations
                        .get(&PropertyId::Filter(VendorPrefix::None))
                        .is_some()
                },
            )
        })
}

#[test]
#[allow(clippy::too_many_lines)]
fn collapse_groups() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove both useless `g`s -->
    <g>
        <g>
            <path d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should pass all inheritable attributes to children -->
    <g>
        <g attr1="val1">
            <path d="..."/>
        </g>
    </g>
    <g attr1="val1">
        <g attr2="val2">
            <path d="..."/>
        </g>
    </g>
    <g attr1="val1">
        <g>
            <path d="..."/>
        </g>
        <path d="..."/>
    </g>
    <g attr1="val1">
        <g attr2="val2">
            <path d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove inheritable overridden attributes -->
    <g attr1="val1">
        <g fill="red">
            <path fill="green" d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove group with equal attribute values to child -->
    <g attr1="val1">
        <g attr2="val2">
            <path attr2="val2" d="..."/>
        </g>
        <g attr2="val2">
            <path attr2="val3" d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should join transform attributes into `transform="rotate(45) scale(2)"` -->
    <g attr1="val1">
        <g transform="rotate(45)">
            <path transform="scale(2)" d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `clip-path` -->
    <clipPath id="a">
       <path d="..."/>
    </clipPath>
    <clipPath id="b">
       <path d="..."/>
    </clipPath>
    <g transform="matrix(0 -1.25 -1.25 0 100 100)" clip-path="url(#a)">
        <g transform="scale(.2)">
            <path d="..."/>
            <path d="..."/>
        </g>
    </g>
    <g transform="matrix(0 -1.25 -1.25 0 100 100)" clip-path="url(#a)">
        <g transform="scale(.2)">
            <g>
                <g clip-path="url(#b)">
                    <path d="..."/>
                    <path d="..."/>
                </g>
            </g>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `clip-path` and `mask` -->
    <clipPath id="a">
       <path d="..."/>
    </clipPath>
    <path d="..."/>
    <g clip-path="url(#a)">
        <path d="..." transform="scale(.2)"/>
    </g>
    <g mask="url(#a)">
        <path d="..." transform="scale(.2)"/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `id` or animation children -->
    <g stroke="#000">
        <g id="star">
            <path id="bar" d="..."/>
        </g>
    </g>
    <g>
        <animate id="frame0" attributeName="visibility" values="visible" dur="33ms" begin="0s;frame27.end"/>
        <path d="..." fill="#272727"/>
        <path d="..." fill="#404040"/>
        <path d="..." fill="#2d2d2d"/>
    </g>
    <g transform="rotate(-90 25 0)">
        <circle stroke-dasharray="110" r="20" stroke="#10cfbd" fill="none" stroke-width="3" stroke-linecap="round">
            <animate attributeName="stroke-dashoffset" values="360;140" dur="2.2s" keyTimes="0;1" calcMode="spline" fill="freeze" keySplines="0.41,0.314,0.8,0.54" repeatCount="indefinite" begin="0"/>
            <animateTransform attributeName="transform" type="rotate" values="0;274;360" keyTimes="0;0.74;1" calcMode="linear" dur="2.2s" repeatCount="indefinite" begin="0"/>
        </circle>
    </g>
</svg>"##
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with classes -->
    <style>
        .n{display:none}
        .i{display:inline}
    </style>
    <g id="a">
        <g class="i"/>
    </g>
    <g id="b" class="n">
        <g class="i"/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve children of `<switch>` -->
    <switch>
        <g id="a">
            <g class="i"/>
        </g>
        <g id="b" class="n">
            <g class="i"/>
        </g>
        <g>
            <g/>
        </g>
    </switch>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should replace inheritable value -->
	<g color="red">
		<g color="inherit" fill="none" stroke="none">
			<circle cx="130" cy="80" r="60" fill="currentColor"/>
			<circle cx="350" cy="80" r="60" stroke="currentColor" stroke-width="4"/>
		</g>
	</g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove useless group -->
    <g filter="url(#...)">
        <g>
            <path d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 88 88">
  <!-- Should preserve group if some attrs cannot be moved -->
  <filter id="a">
    <feGaussianBlur stdDeviation="1"/>
  </filter>
  <g transform="matrix(0.6875,0,0,0.6875,20.34375,66.34375)" style="filter:url(#a)">
    <path d="M 33.346591,-83.471591 L -10.744318,-36.471591 L -10.49989,-32.5" style="fill-opacity:1"/>
  </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
    <!-- Should preserve group if parent has `filter` -->
    <clipPath id="a">
        <circle cx="25" cy="15" r="10"/>
    </clipPath>
    <filter id="b">
        <feColorMatrix type="saturate"/>
    </filter>
    <g filter="url(#b)">
        <g clip-path="url(#a)">
            <circle cx="30" cy="10" r="10" fill="yellow" id="c1"/>
        </g>
    </g>
    <g style="filter:url(#b)">
        <g clip-path="url(#a)">
            <circle cx="20" cy="10" r="10" fill="blue" id="c2"/>
        </g>
    </g>
    <circle cx="25" cy="15" r="10" stroke="black" stroke-width=".1" fill="none"/>
</svg>"#
        )
    )?);

    Ok(())
}
