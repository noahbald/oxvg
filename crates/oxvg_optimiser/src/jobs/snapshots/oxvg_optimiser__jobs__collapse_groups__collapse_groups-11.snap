---
source: crates/oxvg_optimiser/src/jobs/collapse_groups.rs
expression: "test_config(r#\"{ \"collapseGroups\": true }\"#,\n        Some(r#\"<svg xmlns=\"http://www.w3.org/2000/svg\">\n    <!-- Should replace inheritable value -->\n\t<g color=\"red\">\n\t\t<g color=\"inherit\" fill=\"none\" stroke=\"none\">\n\t\t\t<circle cx=\"130\" cy=\"80\" r=\"60\" fill=\"currentColor\"/>\n\t\t\t<circle cx=\"350\" cy=\"80\" r=\"60\" stroke=\"currentColor\" stroke-width=\"4\"/>\n\t\t</g>\n\t</g>\n</svg>\"#))?"
---
<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should replace inheritable value -->
	
	<g color="red" fill="none" stroke="none">
			<circle cx="130" cy="80" r="60" fill="currentColor"></circle>
			<circle cx="350" cy="80" r="60" stroke="currentColor" stroke-width="4"></circle>
		</g>
		
</svg>
